File: app_android/app/src/main/java/com/ridcheck/core/Health.kt
package com.ridcheck.core

import java.util.Locale

/**
 * 健康判定。逐行移植自 app/rid/health.py：
 * - assessPacket：单包判定
 * - StreamAssessor：10s 窗口的速率/停滞/冻结判定
 */
object Health {
    val UAS_ID_CHARSET = Regex("^[0-9A-Z]+$")
    val UAS_ID_FORBIDDEN = setOf('O', 'I')

    const val GB_MIN_RATE_HZ = 1.0 // GB 46750-2025 5.1.3: 广播间隔 ≤ 1s
    const val FRESH_THRESHOLD_S = 2.0 // DATA_FRESH_THRESHOLD_MS = 2000

    /** 判定单个解码包，返回问题列表（空 = 健康）。 */
    fun assessPacket(pkt: DecodedPacket): List<HealthIssue> {
        val issues = ArrayList<HealthIssue>()

        // --- 结构（5.2.3 表3 数据格式） ---
        if (pkt.dataType != Decoder.GB46750_DATA_TYPE) {
            issues.add(HealthIssue(HealthLevel.FAIL, "STRUCT_TYPE",
                String.format("dataType=0x%02X，应为 0xFF", pkt.dataType),
                clause = "5.2.3 表3 数据格式"))
        }
        if (pkt.version != Decoder.GB46750_VERSION) {
            issues.add(HealthIssue(HealthLevel.WARN, "STRUCT_VER",
                String.format("版本=0x%02X，标准为 0x20 (V1.0)；旧版/非标设备降级为警告", pkt.version),
                clause = "5.2.3 表3 数据格式"))
        }
        if (pkt.structureError.isNotEmpty()) {
            issues.add(HealthIssue(HealthLevel.FAIL, "STRUCT_LEN", pkt.structureError,
                clause = "5.2.3 表3 数据格式"))
        }

        // --- UAS_ID (001, M) ---
        val uas = pkt.uasId
        when {
            uas.isEmpty() ->
                issues.add(HealthIssue(HealthLevel.FAIL, "UAS_EMPTY", "唯一产品识别码为空",
                    clause = "表3-001"))
            uas.length != 20 ->
                issues.add(HealthIssue(HealthLevel.WARN, "UAS_LEN",
                    "唯一产品识别码长度 ${uas.length}，应为 20",
                    clause = "表3-001"))
            else -> {
                if (!UAS_ID_CHARSET.matches(uas)) {
                    issues.add(HealthIssue(HealthLevel.WARN, "UAS_CHARSET",
                        "唯一产品识别码含非 0-9/A-Z 字符",
                        clause = "表3-001"))
                }
                if (uas.any { it in UAS_ID_FORBIDDEN }) {
                    issues.add(HealthIssue(HealthLevel.WARN, "UAS_OI",
                        "唯一产品识别码含禁用字符 O/I",
                        clause = "表3-001"))
                }
            }
        }

        // --- 实名登记 (002, M) ---
        if (pkt.realname == "00000000" || pkt.realname.isEmpty()) {
            issues.add(HealthIssue(HealthLevel.WARN, "REALNAME_EMPTY",
                "实名登记号为默认值/空，未完成 UOM 实名登记",
                clause = "表3-002"))
        } else if (pkt.realname.length != 8) {
            issues.add(HealthIssue(HealthLevel.WARN, "REALNAME_LEN",
                "实名登记号长度 ${pkt.realname.length}，应为 8",
                clause = "表3-002"))
        }

        // --- 运行类别/无人机分类 (M) ---
        if (pkt.opCategory !in setOf(1, 2, 3)) {
            issues.add(HealthIssue(HealthLevel.WARN, "OP_CATEGORY",
                "运行类别 ${pkt.opCategory} 无效（应为 1/2/3）",
                clause = "表3-003"))
        }
        if (pkt.uaClass !in setOf(0, 1, 2, 3, 4)) {
            issues.add(HealthIssue(HealthLevel.WARN, "UA_CLASS",
                "无人机分类 ${pkt.uaClass} 无效（应为 0-4）",
                clause = "表3-004"))
        }

        // --- 位置 (008, M) ---
        if (pkt.uaLat.isNaN() || pkt.uaLon.isNaN()) {
            issues.add(HealthIssue(HealthLevel.WARN, "POS_UNKNOWN",
                "无人机位置未知（0xFFFFFFFF）— 未收到 GPS/飞控数据",
                clause = "表3-008"))
        } else {
            if (!(-90.0 <= pkt.uaLat && pkt.uaLat <= 90.0)) {
                issues.add(HealthIssue(HealthLevel.FAIL, "LAT_RANGE",
                    String.format(Locale.US, "纬度 %.7f 超出 [-90,90]", pkt.uaLat),
                    clause = "表3-008"))
            }
            if (!(-180.0 <= pkt.uaLon && pkt.uaLon <= 180.0)) {
                issues.add(HealthIssue(HealthLevel.FAIL, "LON_RANGE",
                    String.format(Locale.US, "经度 %.7f 超出 [-180,180]", pkt.uaLon),
                    clause = "表3-008"))
            }
        }

        // --- 高度 (013, M) ---
        if (pkt.geoAlt.isNaN()) {
            issues.add(HealthIssue(HealthLevel.WARN, "GEOALT_UNKNOWN",
                "大地高度未知（0）— 未收到高度数据",
                clause = "表3-013"))
        } else if (!(-1000.0 <= pkt.geoAlt && pkt.geoAlt <= 10000.0)) {
            issues.add(HealthIssue(HealthLevel.FAIL, "GEOALT_RANGE",
                String.format(Locale.US, "大地高度 %.1fm 超出 [-1000,10000]", pkt.geoAlt),
                clause = "表3-013"))
        }

        // --- 航迹/地速 (009/010, M) ---
        if (pkt.heading.isNaN()) {
            issues.add(HealthIssue(HealthLevel.WARN, "HEADING_UNKNOWN", "航迹角未知（0xFFFF）",
                clause = "表3-009"))
        } else if (!(0.0 <= pkt.heading && pkt.heading < 360.0)) {
            issues.add(HealthIssue(HealthLevel.FAIL, "HEADING_RANGE",
                String.format(Locale.US, "航迹角 %.1f° 超出 [0,360)", pkt.heading),
                clause = "表3-009"))
        }
        if (pkt.speed.isNaN()) {
            issues.add(HealthIssue(HealthLevel.WARN, "SPEED_UNKNOWN", "地速未知（0xFFFF）",
                clause = "表3-010"))
        } else if (!(0.0 <= pkt.speed && pkt.speed <= 6553.5)) {
            issues.add(HealthIssue(HealthLevel.FAIL, "SPEED_RANGE",
                String.format(Locale.US, "地速 %.1fm/s 超出范围", pkt.speed),
                clause = "表3-010"))
        }

        // --- 运行状态 (015, M) ---
        if (pkt.opStatus !in Decoder.OP_STATUS.keys) {
            issues.add(HealthIssue(HealthLevel.FAIL, "STATUS_INVALID",
                "运行状态 ${pkt.opStatus} 无效（应为 0-5）",
                clause = "表3-015"))
        } else if (pkt.opStatus == 4 || pkt.opStatus == 5) {
            issues.add(HealthIssue(HealthLevel.FAIL, "STATUS_FAIL",
                "设备上报故障状态：${Decoder.OP_STATUS[pkt.opStatus]}",
                clause = "表3-015"))
        }

        // --- 遥控站位置 (006, M) ---
        if (pkt.opLat.isNaN() || pkt.opLon.isNaN()) {
            issues.add(HealthIssue(HealthLevel.WARN, "OPPOS_UNKNOWN",
                "遥控站位置未知 — 未配置起飞点/Home",
                clause = "表3-006"))
        } else if (!(-90.0 <= pkt.opLat && pkt.opLat <= 90.0 && -180.0 <= pkt.opLon && pkt.opLon <= 180.0)) {
            issues.add(HealthIssue(HealthLevel.FAIL, "OPPOS_RANGE", "遥控站位置超出合理范围",
                clause = "表3-006"))
        }

        // --- 时间戳 (020, M) ---
        if (pkt.timestampMs == 0L) {
            issues.add(HealthIssue(HealthLevel.WARN, "TS_UNSYNCED", "时间戳为 0 — 设备尚未 GPS 授时",
                clause = "表3-020"))
        } else {
            val now = System.currentTimeMillis()
            val drift = Math.abs(now - pkt.timestampMs)
            if (drift > 5 * 60 * 1000) {
                issues.add(HealthIssue(HealthLevel.WARN, "TS_DRIFT",
                    String.format(Locale.US, "时间戳与手机时钟偏差 %.0fs，GPS 授时可能异常", drift / 1000.0),
                    clause = "表3-020"))
            }
        }

        return issues
    }

    /** 问题 code → 一句整改建议（报告「建议」列 / 文本报告使用）。未知 code 返回空串。 */
    fun advice(code: String): String = when (code) {
        "STRUCT_TYPE" -> "数据包头部 dataType 非 0xFF，非标准运行识别包；核对广播载荷编码。"
        "STRUCT_VER" -> "数据包版本非 0x20（V1.0），可能为旧版/非标固件；建议升级固件。"
        "STRUCT_LEN" -> "数据包长度与头部声明不一致，疑粘包/截断；抓包复测并核对编码长度。"
        "UAS_EMPTY" -> "唯一产品识别码为空；在固件 config 填入 UOM 注册的 20 位 ID。"
        "UAS_LEN" -> "唯一产品识别码长度不符，应为 20 位；核对固件配置。"
        "UAS_CHARSET" -> "识别码含非 0-9/A-Z 字符；仅允许大写字母与数字。"
        "UAS_OI" -> "识别码含禁用字符 O/I（易与 0/1 混淆）；更换为合法字符。"
        "REALNAME_EMPTY" -> "实名登记号为默认值，未完成 UOM 实名登记；登记后填入 8 位登记号。"
        "REALNAME_LEN" -> "实名登记号长度不符，UOM 登记号固定 8 位。"
        "OP_CATEGORY" -> "运行类别无效；应填 1(开放类)/2(特定类)/3(审定类)。"
        "UA_CLASS" -> "无人机分类无效；应填 0(微型)/1(轻型)/2(小型)/3(中型)/4(大型)。"
        "POS_UNKNOWN" -> "无人机位置未知，未收到 GPS/飞控位置；检查 GPS 定位与飞控数据链路。"
        "LAT_RANGE" -> "纬度超出 [-90,90]，数据异常；核对飞控经纬度换算。"
        "LON_RANGE" -> "经度超出 [-180,180]，数据异常；核对飞控经纬度换算。"
        "GEOALT_UNKNOWN" -> "大地高度未上报；检查 GPS 定位或高度数据链路。"
        "GEOALT_RANGE" -> "大地高度超出合理范围；核对高度数据源。"
        "HEADING_UNKNOWN" -> "航迹角未上报；检查飞控航向输出。"
        "HEADING_RANGE" -> "航迹角超出 [0,360)，数据异常；核对航向编码。"
        "SPEED_UNKNOWN" -> "地速未上报；检查飞控速度输出。"
        "SPEED_RANGE" -> "地速超出合理范围；核对速度编码。"
        "STATUS_INVALID" -> "运行状态值无效；核对 GB 状态字节 0-5。"
        "STATUS_FAIL" -> "设备上报识别发送功能失效；按故障排查（天线/功率/模块自检），必要时重启模块。"
        "OPPOS_UNKNOWN" -> "遥控站位置未上报；确认已设置起飞点/Home 或遥控站 GPS 正常。"
        "OPPOS_RANGE" -> "遥控站位置超出合理范围；核对操作员位置数据。"
        "TS_UNSYNCED" -> "时间戳为 0，设备未 GPS 授时；确保模块在有卫星信号处上电。"
        "TS_DRIFT" -> "时间戳与手机时钟偏差大，GPS 授时异常；重启并等待授时收敛。"
        "RATE_ZERO" -> "近 10s 无数据包，广播中断；检查模块电源/天线/飞控数据源。"
        "RATE_LOW" -> "广播速率低于 1 包/s 标准；检查广播间隔配置与射频拥塞。"
        "RATE_SLOW" -> "广播速率接近下限；适当缩短广播间隔或减小干扰。"
        "STALE" -> "数据长时间未更新，链路停滞；检查飞控连接与模块运行状态。"
        "FROZEN" -> "广播内容长时间不变，数据源可能冻结；检查飞控数据输出与模块缓存。"
        else -> ""
    }
}

/**
 * 累积去重后的数据包并产出流式判定。
 * 时钟可注入（nowFunc），便于纯 JVM 单元测试。
 */
class StreamAssessor(
    private val windowS: Double = 10.0,
    private val nowFunc: () -> Long = { System.nanoTime() / 1_000_000 }
) {
    private val times = ArrayDeque<Long>()
    private var lastPkt: DecodedPacket? = null
    private var lastRaw: ByteArray = ByteArray(0)
    private var identSeq = 0
    private var startMs: Long? = null

    fun push(pkt: DecodedPacket) {
        val now = nowFunc()
        if (startMs == null) startMs = now
        val cutoff = now - (windowS * 1000).toLong()
        while (times.isNotEmpty() && times.first() < cutoff) {
            times.removeFirst()
        }
        times.addLast(now)

        if (pkt.raw.contentEquals(lastRaw)) {
            identSeq += 1
        } else {
            identSeq = 1
        }
        lastRaw = pkt.raw.copyOf()
        lastPkt = pkt
    }

    fun reset() {
        times.clear()
        lastPkt = null
        lastRaw = ByteArray(0)
        identSeq = 0
        startMs = null
    }

    fun report(): HealthReport {
        val now = nowFunc()
        val rep = HealthReport()
        val last = lastPkt ?: run {
            rep.level = HealthLevel.FAIL
            rep.note = "尚未收到任何数据包"
            return rep
        }

        val elapsedS = startMs?.let { (now - it) / 1000.0 } ?: 0.0
        if (times.isNotEmpty()) {
            rep.avgRateHz = times.size / Math.min(elapsedS, windowS)
        }
        rep.packetsSeen = times.size
        rep.staleSeconds = (now - times.last()) / 1000.0

        // --- 最新包的单包问题 ---
        val pktIssues = Health.assessPacket(last)
        rep.issues.addAll(pktIssues)
        rep.packetsOk = if (pktIssues.isEmpty()) 1 else 0

        // --- 速率 (GB 5.1.3: ≥ 1/s) ---
        if (elapsedS >= 3.0) {
            if (rep.packetsSeen == 0) {
                rep.issues.add(HealthIssue(HealthLevel.FAIL, "RATE_ZERO",
                    "近 ${windowS.toInt()}s 内未收到数据包",
                    clause = "5.1.3 广播间隔≤1s"))
            } else if (rep.avgRateHz < Health.GB_MIN_RATE_HZ * 0.5) {
                rep.issues.add(HealthIssue(HealthLevel.WARN, "RATE_LOW",
                    String.format(Locale.US, "广播速率仅 %.1f 包/s，低于 1 包/s 标准", rep.avgRateHz),
                    clause = "5.1.3 广播间隔≤1s"))
            } else if (rep.avgRateHz < Health.GB_MIN_RATE_HZ) {
                rep.issues.add(HealthIssue(HealthLevel.WARN, "RATE_SLOW",
                    String.format(Locale.US, "广播速率 %.1f 包/s，略低于 1 包/s", rep.avgRateHz),
                    clause = "5.1.3 广播间隔≤1s"))
            }
        }

        // --- 停滞 (5.1.2 全程连续广播) ---
        if (rep.staleSeconds > 5.0) {
            rep.issues.add(HealthIssue(HealthLevel.FAIL, "STALE",
                String.format(Locale.US, "已 %.0fs 无新数据", rep.staleSeconds),
                clause = "5.1.2 全程连续广播"))
        } else if (rep.staleSeconds > Health.FRESH_THRESHOLD_S) {
            rep.issues.add(HealthIssue(HealthLevel.WARN, "STALE",
                String.format(Locale.US, "数据已 %.0fs 未更新", rep.staleSeconds),
                clause = "5.1.2 全程连续广播"))
        }

        // --- 内容冻结（数据源停滞，而非射频静默；5.1.2 全程连续广播） ---
        if (identSeq >= 5 && last.timestampMs != 0L) {
            rep.issues.add(HealthIssue(HealthLevel.WARN, "FROZEN",
                "广播内容长时间未变化 — 数据源可能停滞",
                clause = "5.1.2 全程连续广播"))
        }

        // --- 判定 = 最严重问题 ---
        var worst = HealthLevel.PASS
        for (i in rep.issues) {
            if (i.level.ordinal > worst.ordinal) worst = i.level
        }
        rep.level = worst

        rep.note = when (rep.level) {
            HealthLevel.PASS -> "设备工作正常，广播符合 GB 46750-2025 要求"
            HealthLevel.WARN -> "存在可改善项，不影响基本广播"
            else -> "存在故障，请根据下方问题清单排查"
        }

        return rep
    }
}

File: app/rid/health.py
"""Built-in judge: tells whether the RID module / its broadcast is healthy.

Per-packet checks mirror the firmware's own validation
(`gb46750_validateFlightData` in main/protocol/rid_messages.cpp) plus
protocol-structure checks and GB 46750-2025 compliance rules (5.1.3 rate,
field presence, real-name registration, sentinel handling).

Stream-level checks (rate, staleness, frozen content) live in
`StreamAssessor`.
"""
from __future__ import annotations

import re
import time
from collections import deque
from typing import Callable

from .decoder import (
    GB46750_DATA_TYPE,
    GB46750_VERSION,
    OP_STATUS,
    SENT_SPEED_HEADING,
    SENT_VSPEED,
)
from .models import DecodedPacket, HealthIssue, HealthLevel, HealthReport

UAS_ID_CHARSET = re.compile(r"^[0-9A-Z]+$")
UAS_ID_FORBIDDEN = set("OI")  # 字符范围: 0-9 及除 O/I 外的大写字母

GB_MIN_RATE_HZ = 1.0     # GB 46750-2025 5.1.3: 广播间隔 ≤ 1s
FRESH_THRESHOLD_S = 2.0  # DATA_FRESH_THRESHOLD_MS = 2000


def _now_ms() -> int:
    return int(time.monotonic() * 1000)


def _is_nan(v: float) -> bool:
    return v != v


def assess_packet(pkt: DecodedPacket) -> list[HealthIssue]:
    """Judging a single decoded packet. Returns issues (empty = healthy)."""
    issues: list[HealthIssue] = []

    # --- structure ---
    if pkt.data_type != GB46750_DATA_TYPE:
        issues.append(HealthIssue(HealthLevel.FAIL, "STRUCT_TYPE",
                                  f"dataType=0x{pkt.data_type:02X}，应为 0xFF"))
    if pkt.version != GB46750_VERSION:
        issues.append(HealthIssue(HealthLevel.FAIL, "STRUCT_VER",
                                  f"版本=0x{pkt.version:02X}，应为 0x20 (V1.0)"))
    if pkt.structure_error:
        issues.append(HealthIssue(HealthLevel.FAIL, "STRUCT_LEN", pkt.structure_error))

    # --- UAS_ID (001, M) ---
    uas = pkt.uas_id
    if not uas:
        issues.append(HealthIssue(HealthLevel.FAIL, "UAS_EMPTY", "唯一产品识别码为空"))
    elif len(uas) != 20:
        issues.append(HealthIssue(HealthLevel.WARN, "UAS_LEN",
                                  f"唯一产品识别码长度 {len(uas)}，应为 20"))
    else:
        if not UAS_ID_CHARSET.match(uas):
            issues.append(HealthIssue(HealthLevel.WARN, "UAS_CHARSET",
                                      "唯一产品识别码含非 0-9/A-Z 字符"))
        if any(ch in UAS_ID_FORBIDDEN for ch in uas):
            issues.append(HealthIssue(HealthLevel.WARN, "UAS_OI",
                                      "唯一产品识别码含禁用字符 O/I"))

    # --- 实名登记 (002, M) ---
    if pkt.realname == "00000000" or not pkt.realname:
        issues.append(HealthIssue(HealthLevel.WARN, "REALNAME_EMPTY",
                                  "实名登记号为默认值/空，未完成 UOM 实名登记"))
    elif len(pkt.realname) != 8:
        issues.append(HealthIssue(HealthLevel.WARN, "REALNAME_LEN",
                                  f"实名登记号长度 {len(pkt.realname)}，应为 8"))

    # --- 运行类别/无人机分类 (M) ---
    if pkt.op_category not in (1, 2, 3):
        issues.append(HealthIssue(HealthLevel.WARN, "OP_CATEGORY",
                                  f"运行类别 {pkt.op_category} 无效（应为 1/2/3）"))
    if pkt.ua_class not in (0, 1, 2, 3, 4):
        issues.append(HealthIssue(HealthLevel.WARN, "UA_CLASS",
                                  f"无人机分类 {pkt.ua_class} 无效（应为 0-4）"))

    # --- 位置 (008, M) ---
    if _is_nan(pkt.ua_lat) or _is_nan(pkt.ua_lon):
        issues.append(HealthIssue(HealthLevel.WARN, "POS_UNKNOWN",
                                  "无人机位置未知（0xFFFFFFFF）— 未收到 GPS/飞控数据"))
    else:
        if not (-90.0 <= pkt.ua_lat <= 90.0):
            issues.append(HealthIssue(HealthLevel.FAIL, "LAT_RANGE",
                                      f"纬度 {pkt.ua_lat:.7f} 超出 [-90,90]"))
        if not (-180.0 <= pkt.ua_lon <= 180.0):
            issues.append(HealthIssue(HealthLevel.FAIL, "LON_RANGE",
                                      f"经度 {pkt.ua_lon:.7f} 超出 [-180,180]"))

    # --- 高度 (013, M) ---
    if _is_nan(pkt.geo_alt):
        issues.append(HealthIssue(HealthLevel.WARN, "GEOALT_UNKNOWN",
                                  "大地高度未知（0）— 未收到高度数据"))
    elif not (-1000.0 <= pkt.geo_alt <= 10000.0):
        issues.append(HealthIssue(HealthLevel.FAIL, "GEOALT_RANGE",
                                  f"大地高度 {pkt.geo_alt:.1f}m 超出 [-1000,10000]"))

    # --- 航迹/地速 (009/010, M) ---
    if _is_nan(pkt.heading):
        issues.append(HealthIssue(HealthLevel.WARN, "HEADING_UNKNOWN",
                                  "航迹角未知（0xFFFF）"))
    elif not (0.0 <= pkt.heading < 360.0):
        issues.append(HealthIssue(HealthLevel.FAIL, "HEADING_RANGE",
                                  f"航迹角 {pkt.heading:.1f}° 超出 [0,360)"))
    if _is_nan(pkt.speed):
        issues.append(HealthIssue(HealthLevel.WARN, "SPEED_UNKNOWN",
                                  "地速未知（0xFFFF）"))
    elif not (0.0 <= pkt.speed <= 6553.5):
        issues.append(HealthIssue(HealthLevel.FAIL, "SPEED_RANGE",
                                  f"地速 {pkt.speed:.1f}m/s 超出范围"))

    # --- 运行状态 (015, M) ---
    if pkt.op_status not in OP_STATUS:
        issues.append(HealthIssue(HealthLevel.FAIL, "STATUS_INVALID",
                                  f"运行状态 {pkt.op_status} 无效（应为 0-5）"))
    elif pkt.op_status in (4, 5):
        issues.append(HealthIssue(HealthLevel.FAIL, "STATUS_FAIL",
                                  f"模块上报故障状态：{OP_STATUS[pkt.op_status]}"))

    # --- 遥控站位置 (006, M) ---
    if _is_nan(pkt.op_lat) or _is_nan(pkt.op_lon):
        issues.append(HealthIssue(HealthLevel.WARN, "OPPOS_UNKNOWN",
                                  "遥控站位置未知 — 未配置起飞点/Home"))
    elif not (-90.0 <= pkt.op_lat <= 90.0 and -180.0 <= pkt.op_lon <= 180.0):
        issues.append(HealthIssue(HealthLevel.FAIL, "OPPOS_RANGE",
                                  "遥控站位置超出合理范围"))

    # --- 时间戳 (020, M) ---
    if pkt.timestamp_ms == 0:
        issues.append(HealthIssue(HealthLevel.WARN, "TS_UNSYNCED",
                                  "时间戳为 0 — 模块尚未 GPS 授时"))
    else:
        now = int(time.time() * 1000)
        drift = abs(now - pkt.timestamp_ms)
        if drift > 5 * 60 * 1000:
            issues.append(HealthIssue(HealthLevel.WARN, "TS_DRIFT",
                                      f"时间戳与电脑时钟偏差 {drift / 1000:.0f}s，GPS 授时可能异常"))

    return issues


class StreamAssessor:
    """Accumulates distinct decoded packets and produces a stream verdict."""

    def __init__(self, window_s: float = 10.0, now_func: Callable[[], int] | None = None):
        self._window_s = window_s
        self._now = now_func or _now_ms
        self._times: deque[int] = deque()
        self._last_pkt: DecodedPacket | None = None
        self._last_raw: bytes = b""
        self._ident_seq = 0
        self._start_ms: int | None = None

    def push(self, pkt: DecodedPacket) -> None:
        now = self._now()
        if self._start_ms is None:
            self._start_ms = now
        cutoff = now - int(self._window_s * 1000)
        while self._times and self._times[0] < cutoff:
            self._times.popleft()
        self._times.append(now)

        if pkt.raw == self._last_raw:
            self._ident_seq += 1
        else:
            self._ident_seq = 1
        self._last_raw = pkt.raw
        self._last_pkt = pkt

    def reset(self) -> None:
        self._times.clear()
        self._last_pkt = None
        self._last_raw = b""
        self._ident_seq = 0
        self._start_ms = None

    def report(self) -> HealthReport:
        now = self._now()
        rep = HealthReport()
        if self._last_pkt is None:
            rep.level = HealthLevel.FAIL
            rep.note = "尚未收到任何数据包"
            return rep

        elapsed_s = (now - self._start_ms) / 1000.0 if self._start_ms else 0.0
        if self._times:
            rep.avg_rate_hz = len(self._times) / min(elapsed_s, self._window_s)
        rep.packets_seen = len(self._times)
        rep.stale_seconds = (now - self._times[-1]) / 1000.0

        # --- packet-level issues from the latest packet ---
        pkt_issues = assess_packet(self._last_pkt)
        rep.issues = list(pkt_issues)
        rep.packets_ok = 1 if not pkt_issues else 0

        # --- rate (GB 5.1.3: ≥ 1/s) ---
        if elapsed_s >= 3.0:
            if rep.packets_seen == 0:
                rep.issues.append(HealthIssue(HealthLevel.FAIL, "RATE_ZERO",
                                              f"近 {self._window_s:.0f}s 内未收到数据包"))
            elif rep.avg_rate_hz < GB_MIN_RATE_HZ * 0.5:
                rep.issues.append(HealthIssue(HealthLevel.WARN, "RATE_LOW",
                                              f"广播速率仅 {rep.avg_rate_hz:.1f} 包/s，低于 1 包/s 标准"))
            elif rep.avg_rate_hz < GB_MIN_RATE_HZ:
                rep.issues.append(HealthIssue(HealthLevel.WARN, "RATE_SLOW",
                                              f"广播速率 {rep.avg_rate_hz:.1f} 包/s，略低于 1 包/s"))

        # --- staleness ---
        if rep.stale_seconds > 5.0:
            rep.issues.append(HealthIssue(HealthLevel.FAIL, "STALE",
                                          f"已 {rep.stale_seconds:.0f}s 无新数据"))
        elif rep.stale_seconds > FRESH_THRESHOLD_S:
            rep.issues.append(HealthIssue(HealthLevel.WARN, "STALE",
                                          f"数据已 {rep.stale_seconds:.0f}s 未更新"))

        # --- frozen content (data source stalled, not just RF silence) ---
        if self._ident_seq >= 5 and self._last_pkt.timestamp_ms != 0:
            rep.issues.append(HealthIssue(HealthLevel.WARN, "FROZEN",
                                          "广播内容长时间未变化 — 数据源可能停滞"))

        # --- verdict = worst issue ---
        worst = max((i.level for i in rep.issues), default=HealthLevel.PASS)
        rep.level = HealthLevel(worst)

        if rep.level == HealthLevel.PASS:
            rep.note = "模块工作正常，广播符合 GB 46750-2025 要求"
        elif rep.level == HealthLevel.WARN:
            rep.note = "存在可改善项，不影响基本广播"
        else:
            rep.note = "存在故障，请根据下方问题清单排查"

        return rep
